plugins/ipam/ipam.go

```diff
@@ -646,6 +645,72 @@ func (i *IPAM) GetIPAMConfigForJSON() *config.IPAMConfig {
+	} else {
+		sort.Slice(serviceIPs, func(i, j int) bool {
+			return bytes.Compare(serviceIPs[i], serviceIPs[j]) < 0
+		})
+		ip = serviceIPs[0].To16()
+	}
+	return i.computeSID(ip, i.ContivConf.GetIPAMConfig().SRv6Settings.ServicePolicyBSIDSubnetCIDR)
+}
+
+// SidForServiceHostLocalsid creates a valid SRv6 SID for service locasid leading to host on the current node. Created SID
+// doesn't depend on anything and is the same for each node, because there is only one way how to get to host in each
+// node and localsid have local significance (their sid don't have to be globally unique)
+func (i *IPAM) SidForServiceHostLocalsid() net.IP {
+	return i.computeSID(net.ParseIP("::1"), i.ContivConf.GetIPAMConfig().SRv6Settings.ServiceHostLocalSIDSubnetCIDR)
+}
+
+// SidForServicePodLocalsid creates a valid SRv6 SID for service locasid leading to pod backend. The SID creation is
+// based on backend IP <backendIP>.
+func (i *IPAM) SidForServicePodLocalsid(backendIP net.IP) net.IP {
+	return i.computeSID(backendIP, i.ContivConf.GetIPAMConfig().SRv6Settings.ServicePodLocalSIDSubnetCIDR)
+}
+
+// SidForNodeToNodePodLocalsid creates a valid SRv6 SID for locasid that is part of node-to-node Srv6 tunnel and
+// outputs packets to pod VRF table.
+func (i *IPAM) SidForNodeToNodePodLocalsid(nodeIP net.IP) net.IP {
+	return i.computeSID(nodeIP, i.ContivConf.GetIPAMConfig().SRv6Settings.NodeToNodePodLocalSIDSubnetCIDR)
+}
+
+// SidForNodeToNodeHostLocalsid creates a valid SRv6 SID for locasid that is part of node-to-node Srv6 tunnel and
+// outputs packets to main VRF table.
+func (i *IPAM) SidForNodeToNodeHostLocalsid(nodeIP net.IP) net.IP {
+	return i.computeSID(nodeIP, i.ContivConf.GetIPAMConfig().SRv6Settings.NodeToNodeHostLocalSIDSubnetCIDR)
+}
+
+// SidForServiceNodeLocalsid creates a valid SRv6 SID for service locasid serving as intermediate step in policy segment list.
+func (i *IPAM) SidForServiceNodeLocalsid(nodeIP net.IP) net.IP {
+	return i.computeSID(nodeIP, i.ContivConf.GetIPAMConfig().SRv6Settings.ServiceNodeLocalSIDSubnetCIDR)
+}
+
+// BsidForNodeToNodePodPolicy creates a valid SRv6 SID for policy that is part of node-to-node Srv6 tunnel and routes traffic to pod VRF table
+func (i *IPAM) BsidForNodeToNodePodPolicy(nodeIP net.IP) net.IP {
+	return i.computeSID(nodeIP, i.ContivConf.GetIPAMConfig().SRv6Settings.NodeToNodePodPolicySIDSubnetCIDR) // bsid = binding sid -> using the same util method
+}
+
+// BsidForNodeToNodeHostPolicy creates a valid SRv6 SID for policy that is part of node-to-node Srv6 tunnel and routes traffic to main VRF table
+func (i *IPAM) BsidForNodeToNodeHostPolicy(nodeIP net.IP) net.IP {
+	return i.computeSID(nodeIP, i.ContivConf.GetIPAMConfig().SRv6Settings.NodeToNodeHostPolicySIDSubnetCIDR) // bsid = binding sid -> using the same util method
+}
+
+// computeSID creates SID by applying network prefix from <prefixNetwork> to IP <ip>
+func (i *IPAM) computeSID(ip net.IP, prefixNetwork *net.IPNet) net.IP {
```

Comment:
maybe there is some go lib that would simplify this function? e.g. https://godoc.org/github.com/apparentlymart/go-cidr/cidr or some other

Reply:
I looked and didn't find anything that could be used for this (even when ignored the goal to simplify code and just use multiple functions from any library to achieve this functionality). 
IMHO, this simple binary arithmetic for ip address is hard to beat in simplicity. Even if some library has exactly what is needed here, the time to understand the library's method can't be much lower then this binary arithmetic logic.